### PR TITLE
[Music] Tidy up CSong and CAlbum  by removing artist vector

### PR DESCRIPTION
**Pull request by DaveTBlake**
Tidy up the way artist data is handled in CSong and CAlbum. Since Gotham we have used a vector of artist credits - an object with both artist name and MusicbrainzID - for artists, but left the legacy string vector duplicating where the names were held. For robustness and consistency the move to using artist credits needs to be completed, and the artist string vector removed.

The changes in this PR depend upon #8079 (CMusicInfoTag MBID setting) due for merge when the window reopens, so those are incuded here. This made methods available that return a string vector from the song or album artist credits as required.

I implementing this change it revealed some places (in JSON interface and album info scraping) where the string vector was being set but the change not propagated to the artist credits, and therefore the database. This is a bug that needs fixing, but I believe a separate PR is best, so for now there are some small bits of code are commented out - they didn't work so no loss!

@evilhamster perhaps you will check this out for me. @zag2me , @jdd-uk you may have guessed this was coming.

### CODE REVIEW

**Comment by evilhamster** [comment]
Have you tested the changes to the CueDocument.cpp with single flac/cue files? To me it looks like this change would lose the albumartist for albums added this way? Since albums are created from sorting FileItems into songs that are later on sorted into albums (see FileItemsToAlbums). The way cue files works to my knowledge is that we create temporary song items that gets converted to fileitems and added to scannedItems and later processed. I think this could be a problem in the json-rpc/AudioLibrary.cpp code as well, but I know to little about how the json interface work. 

Would it be possible to include the bugfixes for the found JSON problems in this PR? 

@Razzeee do you maybe also have some interest in this pr?

**Comment by DaveTBlake** [comment]
No CueDocument.cpp is untested, and yes what is there is not right, I need to do some more work. But what it replaced would not have worked properly either. Anyone got test data for cuesheets? 

In JSON SetSongDetails has bigger problems, the never got propagated to the DB. But rather than add it to this PR, and I am happy to look at it,  I feel it needs to be a separate one as I have doubts about the validity of setting some song details at all. Don't want this improvement to get bogged down with a philosophical discussion of JSON.

**Comment by evilhamster** [comment]
What I used to test the cue functionality was the following:

REM GENRE Reggae
REM DATE 1986
REM DISCID 7C0AD709
REM COMMENT "ExactAudioCopy v0.99pb4"
PERFORMER "Alpha Blondie"
TITLE "Jerusalem"
FILE "Alpha Blondy - Jerusalem.flac" WAVE
  TRACK 01 AUDIO
    TITLE "Jerusalem"
    PERFORMER "Alpha Blondy"
    INDEX 00 00:00:00
    INDEX 01 00:00:33
  TRACK 02 AUDIO
    TITLE "Politiqui"
    PERFORMER "Alpha Blondy"
    INDEX 00 07:49:50
    INDEX 01 07:51:48

And then a flac file, but that one can contain anything. 

I changed the cue so the album artist is different from the track artist. With a earlier 16 alpha I got two artists in the artist table, 
Alpha Blondie
Alpha Blondy

With this pr I didn't get any artist in the artist table, strArtists was empty in the album added (but was fine in the song entries). 

I tested also tested some mixed music (not using cue's but tags for information for those) and strArtists was empty on the album for those as well.

**Comment by jjd-uk** [comment]
I guess this will need @Montellese to check over the JSON stuff.

**Comment by DaveTBlake** [comment]
@Montellese removing the vector has shown up weaknesses in audiolibrary::SetSongDetails and SetAlbumDetails, where changes to artist are not propaged to the song_artist and album_artist tables.  I notice there's a ticket on SetSongDetails because genre isn't propageted properly either.

I am happy to look at fixing that, but would like to do it in a separate PR. The changes here don't break anything that doesn't already not work :)

**Comment by DaveTBlake** [comment]
I think you are right @evilhamster . The caching had me on a goose chase for a while.

>Do you want me to do a pr against your repo or will you implement it? I think it would be best if we avoid doing the same work :)

Thanks, I'm on it now. There is plenty of work to go around without us both doing the same change!!
I'll ping you when there is something to test, if you would be so kind.

**Comment by Razzeee** [comment]
@DaveTBlake 
You might also want to rebase ;)

**Comment by DaveTBlake** [comment]
Have rebased and squished for clarity. 
@evilhamster think I have included everything we discussed consistently. If you are happy I hope we can merge this.

**Comment by evilhamster** [comment]
I think it looks great, but I was able to find one problem. If you do a refresh on a album and have to change the artist name in the dialog, it won't use that name in the next search. To fix that I changed:
 scraper.FindAlbumInfo(album.strAlbum, StringUtils::Join(album.GetAlbumArtist(), g_advancedSettings.m_musicItemSeparator));
to use  album.GetAlbumArtistString() instead of the join statement. 
I also had to update xbmc/music/infoscanner/MusicInfoScanner.cpp  to use:
info.GetAlbum().GetAlbumArtistString() 
album.GetAlbumArtistString()
on lines 1146,1147 instead of the join statements. 

I also found some other places where we should consider use the string version instead of the join variant.

xbmc/addons/Visualisation.cpp 
We have a string artist and albumartist that could be changed. 

xbmc/GUIInfoManager.cpp
A couple of instances where the artiststring version could be used instead of a join on the vector. 

In CSong we now have a SetAlbumArtist, but no SetArtist wouldn't it be more consistent to add a setartist and use that one from CueDocument instead of updating strArtistDesc and having some extra code to handle cuedocument specifics in CMusicInfoTag::SetSong?

**Comment by DaveTBlake** [comment]
Thanks @evilhamster it seems that a few things got lost in my Git manouvres (?), while redoing those I have also extended the use of string method as you suggested rather than all those separate join of vector.

>In CSong we now have a SetAlbumArtist, but no SetArtist wouldn't it be more consistent to add a setartist and use that one from CueDocument instead of updating strArtistDesc and having some extra code to handle cuedocument specifics in CMusicInfoTag::SetSong?

I would rather not do that, trying to avoid a vector interface that could be used by mistake rather than artist credits. CSong has to have an albumartist vector for the album creation process, but once that has happened the data is in album artist credits, hence it is private and a set method needed. CSong does not have an artist vector any more, if we are not using the artist credits then all we have is the artist decription string. Providing a vector form for setting this string could easily be mistaken for setting artist credits and we are back to where we started!

It is not just Cuesheet that doesn't fill the artist credits (also import and JSON, and playing music doesn't either). That is OK, the desc string is enough and CMusicInfoTag::SetSong handles all those situations.

**Comment by zag2me** [comment]
@Razzeee could you have a final look at this. Assuming @evilhamster is OK with the comments you can have my +1

**Comment by jjd-uk** [comment]
As xbmc/interfaces/json-rpc/AudioLibrary.cpp is touched this might need an ok from @Montellese

File: xbmc/music/Album.cpp [hunk 1]
```diff
@@ -85,8 +84,8 @@ CAlbum::CAlbum(const CFileItem& item)
         }
       }
 
-      if (artistName.empty() && tag.GetMusicBrainzAlbumArtistID().size() == artist.size())
-        artistName = artist[i];
+      if (artistName.empty() && tag.GetMusicBrainzAlbumArtistID().size() == tag.GetArtist().size())
+        artistName = tag.GetArtist()[i];
```
Comment (evilhamster) [inline xbmc/music/Album.cpp]: I think this needs to be changed to tag.GetAlbumArtist() instead of tag.GetArtist() for this line and the previous line.

Comment (DaveTBlake) [inline xbmc/music/Album.cpp]: Yes it does, thanks @evilhamster. Fixed

**Comment by evilhamster** [comment]
@DaveTBlake I was thinking that the set function could set the artistcredits for the song, but if you think that this is a better way of doing it that is less error prone that's good enough for me.

All good with this pr from my point of view, looking forward to seeing it merged!

**Comment by Razzeee** [comment]
jenkins build this please

**Comment by DaveTBlake** [comment]
Can someone tell me why the build says check failed? There don't seem to be any fault details. Anything I need to do?

**Comment by stefansaraev** [comment]
linux build error unrelated to this PR.

**Comment by jjd-uk** [comment]
It's not a problem with the PR as far as I can tell, here is job http://jenkins.kodi.tv/job/BuildMulti-PR/3767/

Green = platform is ok
Amber = binary add-on has failed to build
Red = platform has failed to build

So LINUX-64 has had a binary add-on failed to build, click on Build#5750

http://jenkins.kodi.tv/job/LINUX-64/5750/

and you can see screensaver.matrixtrails has failed to be, so not connected to this PR I'd have thought.

**Comment by Razzeee** [comment]
Thanks for taking care of this!

**Comment by ronie** [comment]
this is causing some issues in json-rpc.

when i make a AudioLibrary.GetAlbums call, the returned artist field is an empty list for every album.

ping @Montellese

**Comment by DaveTBlake** [comment]
@ronie I'll take a look first thing since I did ths tidy up. My JSON testing was limited.

**Comment by DaveTBlake** [comment]
I understand what is happening here but I am unsure what it is best to do about it.

AudioLibrary.GetAlbums sees data from the albumview hence the artists are derived from the strArtists field of the album table (not the album_artist table). Since processing of the ARTISTS tag was added this string may differ from the actual individual artists held. It can be used as a description of the artists but just splitting it into a vector does not always give the list of artists that Kodi holds. Hence this PR stopped populating a vector of artists in CMusicInfoTag if the album_artist table had not been queried. The outcome of an empty list for JSON was unintended, but previously we were returning data different from what Kodi was using..

We have two options
a) base AudioLibrary.GetAlbums on a query that does query album_artist table.
b) pass JSON a descrpitive artists string rather than a vector of artists

A third way would be to split the string into artists as before but that will only lead to difficulties if it then tries to get data for these artists, so I advise against that.

If we pick a) then do we want to change albumview (meaning that all calls to GetAlbumsNav use the album_artist table and possibly slow things)? When just displaying the album artists e.g. the library albums node, the descriptive string is enough. 

Your thoughts @evilhamster, @Montellese

**Comment by evilhamster** [comment]
Well before this pr b) was used, so that behavior should probably be returned :)

Some more information, 

The issue is in SetPropertiesFromAlbum where we set two variables:
item.SetProperty("album_artist", album.GetAlbumArtistString());
item.SetProperty("album_artist_array", album.GetAlbumArtist());

The album_artist variable will always contain information that string will be how the artists was credited on the album.

The album_artist_array will only be filled if the albums has loaded the information from the album_artist table. This array will only contain information on the individual artists was credited on the album, but there is no gurantee that an artist in album_artist_array also exists in the artist table (the artist entry in album_artist_array might be an alias). 

Would it be possible to use the property album_artist instead of album_artist_array in what is published over json? If you need individual artists, I think we should give a list of artists db id:s instead so we know it's possible to find any hits for them.

**Comment by DaveTBlake** [comment]
>Well before this pr b) was used, so that behavior should probably be returned.

Not quite accurate. The string is still available, but JSON is looking for data in a vector.

CAlbum had a vector that sometimes contained the artist list from the album_artist table and other times values split from album.strArtists. IMO inconsistent and undesireable. AudioLibrary.GetAlbums used the later, but could also return a vector of artistIds from the album_artist table. The two may not match. But option c) could give matching artist names and artistID

**Comment by Razzeee** [comment]
I think this also causes ```FieldArtist``` to be empty.
Thats the reason why sorting by artist or sorting by artist / year is now broken

**Comment by DaveTBlake** [comment]
Yes you are right, I just spotted it myself. Sorry about that I will fix immediately.

But interesting to note how many things are using a vector of artists that is derrived from the descriptive string of artist names from album or song tables rather than the artists in the song_artist or album_artist tables. They are no longer always the same. 

I will do a sort term fix, but in the longer term I think we need to query the song_artist and album_artist tables
